File: C++/nextGreaterElements.cpp
/**
 * @File    : nextGreaterElements.cpp
 * @Brief   : 下一个更大元素 II
 * @Link    : https://leetcode-cn.com/problems/next-greater-element-ii/
 * @Author  : Wei Li
 * @Date    : 2021-07-27
*/

#include <iostream>
#include <vector>
#include <stack>

class Solution
{
public:
    std::vector<int> nextGreaterElements(std::vector<int> &nums)
    {
        int n = nums.size();
        std::vector<int> ret(n, -1);
        std::stack<int> stk;
        for (int i = 0; i < n * 2 - 1; i++)
        {
            while (!stk.empty() && nums[stk.top()] < nums[i % n])
            {
                ret[stk.top()] = nums[i % n];
                stk.pop();
            }
            stk.push(i % n);
        }
        return ret;
    }
};

template <typename T>
void DisplayElement(const T &container)
{
    for (auto element : container)
    {
        std::cout << element << " ";
    }
    std::cout << std::endl;
}

// -----------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums{1, 2, 1};

    Solution solution;
    auto next_greater_element = solution.nextGreaterElements(nums);
    DisplayElement(next_greater_element);

    return 0;
}

File: Python/nextGreaterElements.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: MyStack.py
@Function: 用队列实现栈
@Link: https://leetcode-cn.com/problems/implement-stack-using-queues/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-27
"""


from typing import List

class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ret = [-1] * n
        stk = list()

        for i in range(n * 2 - 1):
            while stk and nums[stk[-1]] < nums[i % n]:
                ret[stk.pop()] = nums[i % n]
            stk.append(i % n)
        
        return ret


# -------------------------
if __name__ == "__main__":
    nums = [1,2,1]

    solution = Solution()
    next_greater_element = solution.nextGreaterElements(nums)
    print(f"The solution of this problem is\n {next_greater_element}")
